File: src/decentralized_messenger/crypto/keys.py
"""Ed25519 key pair generation, signing, and verification."""

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)


def generate_keypair() -> tuple[bytes, bytes]:
    """Return (private_key_bytes, public_key_bytes) as raw 32-byte values."""
    private_key = Ed25519PrivateKey.generate()
    private_bytes = private_key.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())
    public_bytes = private_key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    return private_bytes, public_bytes


def sign(private_key_bytes: bytes, data: bytes) -> bytes:
    """Sign data with an Ed25519 private key. Returns 64-byte signature."""
    private_key = Ed25519PrivateKey.from_private_bytes(private_key_bytes)
    return private_key.sign(data)


def verify(public_key_bytes: bytes, data: bytes, signature: bytes) -> bool:
    """Return True if signature is valid for data under the given public key."""
    public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes)
    try:
        public_key.verify(signature, data)
        return True
    except Exception:
        return False

File: internal/crypto/keys.go
// Package crypto provides Ed25519 key generation, signing, and verification,
// including helpers that operate directly on SignedMessage values.
package crypto

import (
	"crypto/ed25519"
	"crypto/rand"

	"github.com/Artem-Kornilov-pro/decentralized-messenger/internal/models"
)

// GenerateKeyPair returns a fresh Ed25519 (privateKey, publicKey) pair as raw
// byte slices (64 bytes and 32 bytes respectively).
func GenerateKeyPair() (priv, pub []byte, err error) {
	pubKey, privKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, err
	}
	return privKey, pubKey, nil
}

// Sign signs data with the given raw Ed25519 private key.
func Sign(privateKey, data []byte) []byte {
	return ed25519.Sign(ed25519.PrivateKey(privateKey), data)
}

// Verify reports whether sig is a valid signature of data under publicKey.
func Verify(publicKey, data, sig []byte) bool {
	if len(publicKey) != ed25519.PublicKeySize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(publicKey), data, sig)
}

// SignMessage signs the message's canonical payload and returns a copy with the
// signature attached.
func SignMessage(msg models.SignedMessage, privateKey []byte) models.SignedMessage {
	msg.Signature = Sign(privateKey, msg.SigningPayload())
	return msg
}

// VerifyMessage reports whether the message carries a valid signature under its
// own public key.
func VerifyMessage(msg models.SignedMessage) bool {
	if len(msg.Signature) == 0 {
		return false
	}
	return Verify(msg.PublicKey, msg.SigningPayload(), msg.Signature)
}
